torch/optim/swa_utils.py

```diff
@@ -0,0 +1,267 @@
+        >>> swa_model = torch.optim.swa_utils.AveragedModel(model)
+        >>> scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, 
+        >>>                                     T_max=300)
+        >>> swa_start = 160
+        >>> swa_scheduler = SWALR(optimizer, swa_lr=0.05)
+        >>> for i in range(300):
+        >>>      for input, target in loader:
+        >>>          optimizer.zero_grad()
+        >>>          loss_fn(model(input), target).backward()
+        >>>          optimizer.step()
+        >>>      if i > swa_start:
+        >>>          swa_model.update_parameters(model)
+        >>>          swa_scheduler.step()
+        >>>      else:
+        >>>          scheduler.step()
+        >>>
+        >>> # Update bn statistics for the swa_model at the end
+        >>> torch.optim.swa_utils.update_bn(loader, swa_model) 
+
+    You can use custom averaging functions with `avg_fn` parameter.
+    If no averaging function is provided, the default is to compute
+    equally-weighted average of the weights.
+
+    Example:
+        >>> # ema_model computes exponential moving averages of the weights
+        >>> ema_avg = lambda averaged_model_parameter, model_parameter, num_averaged:\
+                            0.1 * averaged_model_parameter + 0.9 * model_parameter
+        >>> ema_model = torch.optim.swa_utils.AveragedModel(model, avg_fn=ema_avg)
+
+    .. note::
+        When using SWA with models containing Batch Normalization you may 
+        need to update the activation statistics for Batch Normalization.
+        You can do so by using :meth:`torch.optim.swa_utils.update_bn` utility.
+
+    .. note::
+        :attr:`avg_fn` is not saved in the :meth:`state_dict` of the model.
+
+    .. note::
+        When :meth:`update_parameters` is called for the first time (i.e. 
+        :attr:`n_averaged` is `0`) the parameters of `model` are copied
+        to the parameters of :class:`AveragedModel`. For every subsequent
+        call of :meth:`update_parameters` the function `avg_fn` is used
+        to update the parameters.
+
+    .. _Averaging Weights Leads to Wider Optima and Better Generalization:
+        https://arxiv.org/abs/1803.05407
+    .. _There Are Many Consistent Explanations of Unlabeled Data: Why You Should
+        Average:
+        https://arxiv.org/abs/1806.05594
+    .. _SWALP: Stochastic Weight Averaging in Low-Precision Training:
+        https://arxiv.org/abs/1904.11943
+    .. _Stochastic Weight Averaging in Parallel: Large-Batch Training That 
+        Generalizes Well:
+        https://arxiv.org/abs/2001.02312
+    """
+    def __init__(self, model, device=None, avg_fn=None):
+        super(AveragedModel, self).__init__()
+        self.module = deepcopy(model)
+        if device is not None:
+            self.module = self.module.to(device)
+        self.register_buffer('n_averaged',
+                             torch.tensor(0, dtype=torch.long, device=device))
```

Comment:
I would expect `n_averaged` to be created on `self.module.device`, or on default device (i.e. without specifying a `device=` option). Is there a reason why this is not the case?

Reply:
Right now if `device` is specified, we put `n_averaged` on `device`, otherwise it's put on the default device (`cpu`). I can also put `n_averaged` on the `next(iter(model.parameters())).device` in case `device` is not specified if you think that's a better idea. As for `self.module.device`, I think `torch.nn.Module` doesn't have to have a `device` attribute?

Reply:
Yes, what is there is reasonable.